### PR TITLE
Add logmeanexp.

### PR DESCRIPTION
**Pull request by carlosgmartin**
Relanding https://github.com/jax-ml/jax/pull/30206 due to https://github.com/jax-ml/jax/pull/30266.

@jakevdp Could you describe the error in more detail, and how it should be solved? The sum inside `_count` is computed using the specified `dtype`. Inside `_mean`, we pass `dtype=computation_dtype`, just as before, to `_count`. Thus `sum` ultimately should receive `dtype=computation_dtype`, just as before.

### CODE REVIEW

**Comment by jakevdp** [comment]
I don't have a reproducer, but we had downstream reports of the mean ~returning the wrong result~ erroring for large arrays. I suspect it had something to do with the precision of the accumulator in the denominator.

**Comment by carlosgmartin** [comment]
That's strange.

In the [current version](https://github.com/jax-ml/jax/pull/30769/files#diff-2ee33b05046e082a486acddbd6ccfa7568c624035cf6b0787db75952080c2c15L894), `sum` is called with `dtype=computation_dtype`.

In the [new version](https://github.com/jax-ml/jax/pull/30769/files#diff-2ee33b05046e082a486acddbd6ccfa7568c624035cf6b0787db75952080c2c15R880), `sum` is called with `dtype=dtype`, where `dtype` has been [set](https://github.com/jax-ml/jax/pull/30769/files#diff-2ee33b05046e082a486acddbd6ccfa7568c624035cf6b0787db75952080c2c15R910) to `dtype=computation_dtype`.

Thus there shouldn't be any difference.

I get expected results with the new version for large arrays:

```python3
import jax
from jax import nn, numpy as jnp
jax.config.update('jax_enable_x64', True)

nn.logmeanexp  # confirm that the new version is being used
print(jnp.mean(jnp.full([2**31], 999, dtype=int)))  # 999.0
print(jnp.mean(jnp.full([2**31], 2**30, dtype=int)))  # 1073741824.0
print(jnp.mean(jnp.arange(2**31)))  # 1073741823.5
print(2**31 / 2)  # 1073741824.0
```

Maybe it's a platform-specific or hardware-related bug?

**Comment by jakevdp** [comment]
I think the issue is with this line:
```python
count = lax.asarray(count).astype(dtype)
```

If X64 mode is not enabled and `count` is an integer larger than the int32 max, then `lax.asarray(count)` overflows and you get an error. The current implementation avoids this because it calls `lax.convert_element_type(normalizer, computation_dtype)` rather than `lax.asarray(normalizer).astype(computation_dtype)`, and `lax.convert_element_type` has special handling for casting Python ints to float type.

Your test above didn't hit this error because you enabled X64.

**Comment by jakevdp** [comment]
I think we may be able to add a regression test for this without having to allocate gigabytes of data by using `eval_shape`; e.g.
```python
x = jax.ShapeDtypeStruct((1<<32,), jnp.dtype('float32'))
jax.eval_shape(jnp.mean, x)  # doesn't error
```

**Comment by carlosgmartin** [comment]
@jakevdp Good point. Updated.

File: jax/_src/numpy/reductions.py [hunk 1]
```diff
@@ -863,6 +863,25 @@ def mean(a: ArrayLike, axis: Axis = None, dtype: DTypeLike | None = None,
   return _mean(a, _ensure_optional_axes(axis), dtype, out, keepdims,
                where=where, upcast_f16_for_computation=(dtype is None))
 
+def _count(
+    a: ArrayLike,
+    axis: Axis,
+    keepdims: bool=False,
+    where: ArrayLike | None = None,
+    dtype: DTypeLike | None = None,
+):
```
Comment (jakevdp) [inline jax/_src/numpy/reductions.py:872]: It looks like this would wrap the value if `dtype` is an integer and `a.shape[axis]` overflows. Should we raise on that instead? I realize that as currently used, the dtype is always floating point, but if we're including this utility we should make it robust to prevent potential future bugs.

An alternative would be to error if `dtype` is not inexact, with a comment that we should think about overflow if we ever need that code path.

Comment (carlosgmartin) [reply jax/_src/numpy/reductions.py:872]: Is the following what you're thinking?

```python3
if dtypes.issubdtype(dtype, np.inexact):
  count = lax.convert_element_type(count, dtype)
else:
  # raises on overflow
  count = lax.asarray(count).astype(dtype)
```
And what is the expected behavior for `dtype=None`? Should it be returning a plain Python integer, a JAX float, or a JAX int?

Comment (jakevdp) [reply jax/_src/numpy/reductions.py:872]: That would be an option, but you'd probably want to add a test to confirm it overflows where expected.

Another option would be something like
```python
if not dtypes.issubdtype(dtype, np.inexact):
  # before enabling this code path we should think about what to do with overflows
  raise NotImplemenetedError("count() with integer output type")
```

> And what is the expected behavior for `dtype=None`?

This is an internal utility that so far only you are using – you can decide what's most convenient. Maybe don't have any default dtype for now?

Comment (jakevdp) [reply jax/_src/numpy/reductions.py:872]: Maybe a principle here: if you want to make this utility general, you should unit test it to cover all the ways it might be called. If you don't want to write unit tests for it, limit it to the precise cases you use in `mean` and `logmean`.

Comment (carlosgmartin) [reply jax/_src/numpy/reductions.py:872]: [`tests.lax_numpy_reducers_test.JaxNumpyReducerTests.testReducer`](https://github.com/jax-ml/jax/blob/fcc09f53b0caca657560c0d270a33716f1b28f82/tests/lax_numpy_reducers_test.py#L206) calls `jnp.mean` with `dtype=numpy.int16` in one of its test cases, triggering that code path.

It also looks like NumPy wraps around in that case:
```
$ py -c "import numpy as np; print(np.mean(np.ones(2**15), dtype=np.int16))"
-1
$ py -c "import numpy as np; print(np.mean(np.ones(2**16), dtype=np.int16))"
0
```

Comment (jakevdp) [reply jax/_src/numpy/reductions.py:872]: As long as this PR doesn't change how JAX handles integer overflow within the denominator of `jnp.mean`, I think it's probably fine.

Last time, we had to roll back because a corner case of `jnp.mean` was inadvertently changed by the refactor in this PR. We should do everything we can to avoid that this time if we want to avoid another rollback.

Comment (carlosgmartin) [reply jax/_src/numpy/reductions.py:872]: Ok, I've preserved the `lax.convert_element_type` behavior of the original `jnp.mean`.

File: tests/nn_test.py [hunk 2]
```diff
@@ -720,6 +720,30 @@ def f(hx, _):
     with jax.checking_leaks():
       fwd()  # doesn't crash
 
+  @parameterized.product(
+      shape=[(5,), (3, 5), (2, 3, 5)],
+      use_where=[True, False],
+      keepdims=[True, False],
+  )
+  def testLogMeanExp(self, shape, use_where, keepdims):
+    key = random.key(0)
```
Comment (jakevdp) [inline tests/nn_test.py]: Typically for generating input values in tests, we use `self.rng()` and generate via NumPy.

The benefit of this is that the random seed is set via a deterministic hash of the test function name, which avoids potential statistical problems due to repeated use of common seeds like `0` and `42` across the test suite (also for a fully-eager python process like a unit test, there is little benefit to using JAX's stateless RNG).

Could you change the input generation here to use the same strategy?

**Review by jakevdp — APPROVED** [review]
Looks great, thanks!

File: jax/_src/nn/functions.py [hunk 3]
```diff
@@ -518,6 +519,32 @@ def glu(x: ArrayLike, axis: int = -1) -> Array:
...
+def logmeanexp(
+    x: ArrayLike,
+    axis: int | tuple[int, ...] | None = None,
+    where: ArrayLike | None = None,
+    keepdims: bool = False,
+) -> Array:
+  r"""Log mean exp.
+  Computes the function:
+  .. math::
+    \text{logmeanexp}(x) = \log \frac{1}{n} \sum_{i=1}^n \exp x_i = \text{logsumexp}(x) - \log n
```
Comment (jakevdp) [inline jax/_src/nn/functions.py:534]: This is leading to doc build errors:
```
/__w/jax/jax/jax/_src/nn/functions.py:docstring of jax._src.nn.functions.logmeanexp:4: ERROR: Unexpected indentation.
/__w/jax/jax/jax/_src/nn/functions.py:docstring of jax._src.nn.functions.logmeanexp:5: WARNING: Literal block ends without a blank line; unexpected unindent.
```
I think you need a blank line before and after the math block.

File: jax/_src/nn/functions.py [hunk 4]
```diff
@@ -518,6 +519,32 @@ def glu(x: ArrayLike, axis: int = -1) -> Array:
...
 
+@partial(api.jit, static_argnames=("axis", "keepdims"))
+def logmeanexp(
+    x: ArrayLike,
+    axis: int | tuple[int, ...] | None = None,
+    where: ArrayLike | None = None,
+    keepdims: bool = False,
+) -> Array:
+  r"""Log mean exp.
+  Computes the function:
```
Comment (jakevdp) [inline jax/_src/nn/functions.py:531]: Add a blank line above this (otherwise I think it will be treated as a continuation of the title line)

**Review by jakevdp — APPROVED** [review]

**Comment by carlosgmartin** [comment]
@jakevdp Now that this is merged, should we perhaps log a warning when `lax.convert_element_type(count, dtype)` wraps? Even though NumPy silently wraps, this seems like something worth alerting users to.

**Comment by jakevdp** [comment]
Yeah that would be worth looking into: I think it's just a one-line change (but landing it would likely involve fixing a lot of downstream failures that currently pass silently)

**Comment by carlosgmartin** [comment]
What's the one-line change you have in mind?
